conan/tools/microsoft/visual.py

```diff
@@ -275,3 +276,39 @@ def is_msvc_static_runtime(conanfile):
     :return: True, if msvc + runtime MT. Otherwise, False
     """
     return is_msvc(conanfile) and "MT" in msvc_runtime_flag(conanfile)
+
+
+def msvs_toolset(conanfile):
+    """ Returns the corresponding Visual Studio/msvc platform toolset based on the settings of the
+        given conanfile
+    :param conanfile: Conanfile instance
+    :return: Microsoft toolset when compiler is valid. Otherwise, an empty string.
+    """
+    settings = conanfile.settings
+    compiler = settings.get_safe("compiler")
+    compiler_version = settings.get_safe("compiler.version")
+    if compiler == "msvc":
+        subs_toolset = settings.get_safe("compiler.toolset")
+        if subs_toolset:
+            return subs_toolset
+        return msvc_version_to_toolset_version(compiler_version)
+    if compiler == "intel":
+        compiler_version = compiler_version if "." in compiler_version else \
+            "%s.0" % compiler_version
+        return "Intel C++ Compiler " + compiler_version
+    if compiler == "intel-cc":
+        return IntelCC(conanfile).ms_toolset
```

Comment:
seems like also raises in some cases https://github.com/conan-io/conan/blob/bcbaf3958818e7f1d8da87c57a3c8fb793852388/conan/tools/intel/intel_cc.py#L45

Reply:
it's accurate as you are using intel-cc compiler on Darwin. It's a case which I barely can think we will have in production.

Reply:
there are more places where it raises (https://github.com/conan-io/conan/blob/bcbaf3958818e7f1d8da87c57a3c8fb793852388/conan/tools/intel/intel_cc.py#L51)
but my point is, we choose one of two:
- we make an effort that tools doesn't raise, ever, and only returns `None` if it cannot find toolset
- or we document a behavior that function may raise in certain scenarios

Reply:
> or we document a behavior that function may raise in certain scenarios

Sounds reasonable. I'll update the PR https://github.com/conan-io/docs/pull/2951